gcore/geoheif.cpp

```diff
@@ -0,0 +1,195 @@
+{
+    // TODO: this only handles the 2D case.
+    if (length != (6 * 8 + 4))
+    {
+        return;
+    }
+    // Match version
+    if (payload[0] == 0x00)
+    {
+        uint32_t index = 0;
+        if (payload[index + 3] == 0x01)
+        {
+            index += 4;
+            modelTransform[1] = to_double(payload, index);
+            index += 8;
+            modelTransform[2] = to_double(payload, index);
+            index += 8;
+            modelTransform[0] = to_double(payload, index);
+            index += 8;
+            modelTransform[4] = to_double(payload, index);
+            index += 8;
+            modelTransform[5] = to_double(payload, index);
+            index += 8;
+            modelTransform[3] = to_double(payload, index);
+        }
+    }
+}
+
+CPLErr GeoHEIF::GetGeoTransform(double *padfTransform) const
+{
+    padfTransform[1] = modelTransform[1];
+    padfTransform[2] = modelTransform[2];
+    padfTransform[0] = modelTransform[0];
+    padfTransform[4] = modelTransform[4];
+    padfTransform[5] = modelTransform[5];
+    padfTransform[3] = modelTransform[3];
+    return CE_None;
+}
+
+/************************************************************************/
+/*                          GetSpatialRef()                             */
+/************************************************************************/
+const OGRSpatialReference *GeoHEIF::GetSpatialRef() const
+{
+    return &m_oSRS;
+}
+
+void GeoHEIF::extractSRS(const uint8_t *payload, size_t length) const
+{
+    // TODO: more sophisticated length checks
+    if (length < 6)
+    {
+        return;
+    }
+    std::string crsEncoding(payload + 4, payload + 8);
+    std::string crs(payload + 8, payload + length);
+    if (crsEncoding == "wkt2")
+    {
+        m_oSRS.importFromWkt(crs.c_str());
+    }
+    else if (crsEncoding == "crsu")
+    {
+        m_oSRS.importFromCRSURL(crs.c_str());
+    }
+    else if (crsEncoding == "curi")
+    {
+        if ((crs.at(0) != '[') || (crs.at(crs.length() - 1) != ']'))
+        {
+            return;
+        }
+        std::string curie = crs.substr(1, crs.length() - 2);
+        size_t seperatorPos = curie.find(':');
+        std::string authority = curie.substr(0, seperatorPos);
+        std::string code = curie.substr(seperatorPos + 1);
+        std::string osURL("http://www.opengis.net/def/crs/");
+        osURL.append(authority);
+        osURL += "/0/";
+        osURL.append(code);
+        m_oSRS.importFromCRSURL(osURL.c_str());
+    }
+    else
+    {
+        return;
+    }
+}
+
+void GeoHEIF::addGCPs(const uint8_t *data, size_t length)
+{
+    if (data[0] == 0x00)
+    {
+        uint32_t index = 0;
+        bool is_3D = (data[index + 3] == 0x00);
+        index += 4;
+        uint16_t count = (data[index] << 8) + (data[index + 1]);
+        index += 2;
+        for (uint16_t j = 0; (j < count) && (index < length); j++)
+        {
+            GDAL_GCP gcp;
```

Comment:
you will have a leak of memory using a ``std::vector<GDAL_GCP>`` with the C GDAL_GCP if you don't manually free the pszId and pszInfo members. I'd suggest instead you use rather the C++ ``gdal::GCP`` (cf gcore/gdal_priv.h) to have a ``std::vector<gdal::GCP>`` that will provide automatic memory allocation/deallocation
Cf https://github.com/OSGeo/gdal/pull/9285 for the PR that introduces it and also shows how existing code is converted from using GDAL_GCP to gdal::GCP

Reply:
Thanks. That is indeed better. However it would be worth checking that I did it correctly.

In particular, I no longer fill in the GCP id with a counter. Not sure if that will make any difference.